envoy/auth_server/proto/gateway_endpoint.proto

```diff
@@ -0,0 +1,122 @@
+// interact with the remote gRPC server.
+service GatewayEndpoints {
+  // FetchAuthDataSync requests the initial set of GatewayEndpoints from the remote gRPC server.
+  rpc FetchAuthDataSync(AuthDataRequest) returns (AuthDataResponse);
+
+  // StreamAuthDataUpdates listens for updates from the remote gRPC server and streams them to the client.
+  rpc StreamAuthDataUpdates(AuthDataUpdatesRequest) returns (stream AuthDataUpdate);
+}
+
+message AuthDataRequest {}
+
+message AuthDataResponse {
+  map<string, GatewayEndpoint> endpoints = 1;
+}
+
+message AuthDataUpdatesRequest {}
+
+// An AuthDataUpdate message is sent from the remote gRPC server when a GatewayEndpoint is created, updated, or deleted.
+message AuthDataUpdate {
+  // A unique identifier for a user account's GatewayEndpoint.
+  // This is used to identify the GatewayEndpoint when making a service request.
+  // It must be passed as the last path segment of the service request URL.
+  // For example: POST https://api.path.xyz/v1/{gateway_endpoint_id}
+  string endpoint_id = 1;
+
+  // The GatewayEndpoint to upsert to the database.
+  // If delete is true, this field should be empty and the associated endpoint_id will be deleted.
+  GatewayEndpoint gateway_endpoint = 2;
+
+  // Indicates whether the GatewayEndpoint should be deleted.
+  bool delete = 3;
+}
+
+// A GatewayEndpoint represents a user account's endpoint, which has two primary functions:
+//  1. Identifying which endpoint is being used to make a service request.
+//  2. Allowing configuration of endpoint-specific settings, such as API key authorization, etc.
+//
+// A GatewayEndpoint is associated to a single UserAccount.
+// A single UserAccount can have multiple GatewayEndpoints.
+// Settings related to service requests, such as enforcing API key authorization, are configured per GatewayEndpoint.
+message GatewayEndpoint {
+  // The unique identifier for a UserAccount's endpoint.
+  // It must be passed as the last path segment of the service request URL; eg `/v1/{endpoint_id}`.
+  // This is used to identify the GatewayEndpoint when making a service request.
+  string endpoint_id = 1;
+
+  // The authorization settings for the GatewayEndpoint.
+  Auth auth = 2;
+
+  // The rate limiting settings for the GatewayEndpoint.
+  // This includes both throughput (TPS) limit and the capacity (longer period) limit.
+  RateLimiting rate_limiting = 3;
+
+  // Optional metadata for the GatewayEndpoint, which can be set to any additional information.
+  map<string, string> metadata = 4;
+}
+
+// The authorization settings for a GatewayEndpoint.
+message Auth {
+  enum AuthType {
+    NO_AUTH = 0;
+    API_KEY_AUTH = 1;
+    JWT_AUTH = 2;
+  }
+
+  // The type of authentication being used.
+  AuthType auth_type = 1;
+
+  oneof auth_type_details {
+    Empty no_auth = 2;
+    // The API key authorization settings for the GatewayEndpoint.
+    string api_key = 3;
```

Comment:
Can you create a `StaticAPIKey` struct with just a single string in there? 

That way we'll be a lot more forward-compatible in case complexity increases.

Reply:
+1 to this. It also makes reading the proto file easier.

Reply:
Sure, that's how I initially designed it but recently updated for simplicity. I can put it back.